### ISSUE TITLE
Escaping inside codeblocks

### ISSUE THREAD

**Issue by ZeLarpMaster** [pragdave/earmark#307]
[The documentation](https://hexdocs.pm/earmark/Earmark.html#module-security) says
> It is your job to sanitize and or filter the output of `Earmark.as_html` if you cannot trust the input and are to serve the produced HTML on the Web.

But how exactly do we do this?

I tried (as [you can see by yourself here](https://github.com/ZeLarpMaster/EarmarkTest)) escaping, not escaping, considering it raw, and not considering it raw, but none of them look good. Not considering it raw is obviously not what I want so what about the other 2 options?

If I don't escape the markdown first, any code outside triple backticks is vulnerable to XSS scripting and therefore a security hole. You can see this by the beautiful `alert(1)` which pops up when you open the page. Considering it's my job to sanitize, I can understand why this happens and I don't really expect anything different.

However, if I do escape the markdown first, then the code outside triple backticks isn't vulnerable and will display properly, but code *inside* triple backticks will be escaped twice and end up being displayed as escaped html (which you can see if you open the raw HTML source on your browser, it will look like: `<pre><code class="html">&amp;lt;script&amp;gt;alert(&amp;quot;No XSS, but not looking good&amp;quot;);&amp;lt;/script&amp;gt;</code></pre>`).

I tried finding out why this happened and I think it happens [here in the HtmlRenderer](https://github.com/pragdave/earmark/blob/master/lib/earmark/html_renderer.ex#L276), but I'm not sure of the implications of simply changing this boolean.

Are we expected to implement our own renderer for this? What should I do?

**Comment by RobertDober** [comment]
> I tried (as you can see by yourself here) escaping, not escaping, considering it raw, and not 
> considering it raw, but none of them look good. Not considering it raw is obviously not what I 
> want so what about the other 2 options?
Can you provide a link to the corresponding code please?

However, escaping the Markdown is **not** the idea, but sanitizing the generated HTML.

In general, sanitizing HTML is not a trivial task and you should probably use a library.
As a Rails and Phoenix user I just use the frameworks' capabilities so I am not of great help.

If you rather want to modify Earmark's behaviour I suggest you use the AST, which is however still an unstable feature.

**Comment by ZeLarpMaster** [comment]
Here's the link to the project I was using: https://github.com/ZeLarpMaster/EarmarkTest

So if I'm understanding this right, I should be using a lib like [phoenix_html_sanitizer](https://github.com/elixirstatus/phoenix_html_sanitizer) to filter the tags? I don't see it in the docs, but is there a list of tags which can be output by Earmark somewhere or should I read through the source code and make the list myself?

**Comment by RobertDober** [comment]
I am not sure I do understand what docs you are talking about?

What list of tags should there be? All tags are output right now. Maybe the docs talking about the rendering of HTML inside markdown can be improved, PRs are certainly welcome :blush:, but I am not sure.

To my understanding, yes, you should use a tool of some kind, and if that tool has a problem with the HTML generated by Earmark your fallback solution would be to transform the AST provided by Earmark.

If you use the AST feedback will be more than welcome and features are probably to be developed to make the treatment and AST retransformation to HTM experience as useful as possible. 

But

**Comment by ZeLarpMaster** [comment]
I was talking about Earmark's docs. The problem here is that I want to avoid XSS attacks, so I obviously want to filter out script HTML tags and any other kind of unnecessary HTML tags (like iframes). A list of what Earmark can output would be useful as a whitelist for sanitizing.

I think I'll use the AST instead. I was a bit unsure about using it since it's "experimental" and likely to change, but I guess I can use it anyway and give feedback instead.

Thank you for your help and quick response!

**Comment by RobertDober** [comment]
That is great, that is what the AST is for, and the usability of such features depends a lot on early adapters.
Looking forward to your feedback.

**Comment by ZeLarpMaster** [comment]
@jdmarshall I know those things don't make sense. I was trying to make it work at all, not trying to make it work perfectly when I wrote this issue.

I ended up [forking this repo and doing the change I needed](https://github.com/ZeLarpMaster/earmark/commit/35c9661d6647059e507c0278347e21d92351c417). I tried the AST way (I see I have some commits on my project where I had started with that), but ended up giving up for a reason I don't remember (it's been more than a year and I apparently didn't keep a good trace in that project).

That single boolean change is all I needed, I sanitize early, call `Earmark.as_html`, then I make Phoenix consider the html as raw. It works fine for my use case.
